In [12]:
import pandas as pd
pd.set_option('display.max_rows', 100)
import numpy as np
import seaborn as sns

In [6]:
houses = pd.read_csv("train.csv")

In [7]:
houses

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0        1          60       RL         65.0     8450   Pave   NaN      Reg   
1        2          20       RL         80.0     9600   Pave   NaN      Reg   
2        3          60       RL         68.0    11250   Pave   NaN      IR1   
3        4          70       RL         60.0     9550   Pave   NaN      IR1   
4        5          60       RL         84.0    14260   Pave   NaN      IR1   
...    ...         ...      ...          ...      ...    ...   ...      ...   
1455  1456          60       RL         62.0     7917   Pave   NaN      Reg   
1456  1457          20       RL         85.0    13175   Pave   NaN      Reg   
1457  1458          70       RL         66.0     9042   Pave   NaN      Reg   
1458  1459          20       RL         68.0     9717   Pave   NaN      Reg   
1459  1460          20       RL         75.0     9937   Pave   NaN      Reg   

     LandContour Utilities  ... PoolArea PoolQC  Fence MiscFeature MiscVal  \
0            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
2            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
3            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
4            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
...          ...       ...  ...      ...    ...    ...         ...     ...   
1455         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1456         Lvl    AllPub  ...        0    NaN  MnPrv         NaN       0   
1457         Lvl    AllPub  ...        0    NaN  GdPrv        Shed    2500   
1458         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1459         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   

     MoSold YrSold  SaleType  SaleCondition  SalePrice  
0         2   2008        WD         Normal     208500  
1         5   2007        WD         Normal     181500  
2         9   2008        WD         Normal     223500  
3         2   2006        WD        Abnorml     140000  
4        12   2008        WD         Normal     250000  
...     ...    ...       ...            ...        ...  
1455      8   2007        WD         Normal     175000  
1456      2   2010        WD         Normal     210000  
1457      5   2010        WD         Normal     266500  
1458      4   2010        WD         Normal     142125  
1459      6   2008        WD         Normal     147500  

[1460 rows x 81 columns]

# Business problem:
## Who might benefit from this project?  How? 
- Housing companies like Zillow or Redfin, maybe banks, could also help buyers and sellers to know how to predict house values

## What is the goal?
- Model house values give certain inputs
1. Analyze house dataset for actionable insights
2. Create model with .138 or lower log RMSE

# Data description (initial EDA):
## What is the source of the data?
- Kaggle competition dataset

## What does it describe? (What is a row?)
- All the characteristics of a house. Seemingly a row is a house.

## What is the shape of the data?
1. nulls, column types
- The shape of the DataFrame is (1460, 81)

In [58]:
houses.drop_duplicates()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0        1          60       RL         65.0     8450   Pave   NaN      Reg   
1        2          20       RL         80.0     9600   Pave   NaN      Reg   
2        3          60       RL         68.0    11250   Pave   NaN      IR1   
3        4          70       RL         60.0     9550   Pave   NaN      IR1   
4        5          60       RL         84.0    14260   Pave   NaN      IR1   
...    ...         ...      ...          ...      ...    ...   ...      ...   
1455  1456          60       RL         62.0     7917   Pave   NaN      Reg   
1456  1457          20       RL         85.0    13175   Pave   NaN      Reg   
1457  1458          70       RL         66.0     9042   Pave   NaN      Reg   
1458  1459          20       RL         68.0     9717   Pave   NaN      Reg   
1459  1460          20       RL         75.0     9937   Pave   NaN      Reg   

     LandContour Utilities  ... PoolArea PoolQC  Fence MiscFeature MiscVal  \
0            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
2            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
3            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
4            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
...          ...       ...  ...      ...    ...    ...         ...     ...   
1455         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1456         Lvl    AllPub  ...        0    NaN  MnPrv         NaN       0   
1457         Lvl    AllPub  ...        0    NaN  GdPrv        Shed    2500   
1458         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1459         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   

     MoSold YrSold  SaleType  SaleCondition  SalePrice  
0         2   2008        WD         Normal     208500  
1         5   2007        WD         Normal     181500  
2         9   2008        WD         Normal     223500  
3         2   2006        WD        Abnorml     140000  
4        12   2008        WD         Normal     250000  
...     ...    ...       ...            ...        ...  
1455      8   2007        WD         Normal     175000  
1456      2   2010        WD         Normal     210000  
1457      5   2010        WD         Normal     266500  
1458      4   2010        WD         Normal     142125  
1459      6   2008        WD         Normal     147500  

[1460 rows x 81 columns]

In [ ]:
houses["Id"].value_counts().sum() == houses.shape[0] 
# True so Id column is unique per row

np.True_

In [107]:
for _ in houses.describe(exclude="object").columns:
    print(_)

Id
MSSubClass
LotFrontage
LotArea
OverallQual
OverallCond
YearBuilt
YearRemodAdd
MasVnrArea
BsmtFinSF1
BsmtFinSF2
BsmtUnfSF
TotalBsmtSF
1stFlrSF
2ndFlrSF
LowQualFinSF
GrLivArea
BsmtFullBath
BsmtHalfBath
FullBath
HalfBath
BedroomAbvGr
KitchenAbvGr
TotRmsAbvGrd
Fireplaces
GarageYrBlt
GarageCars
GarageArea
WoodDeckSF
OpenPorchSF
EnclosedPorch
3SsnPorch
ScreenPorch
PoolArea
MiscVal
MoSold
YrSold
SalePrice


# Columns
## Index
Id

## Object Columns
MSZoning
Street
Alley
LotShape
LandContour
Utilities
LotConfig
LandSlope
Neighborhood
Condition1
Condition2
BldgType
HouseStyle
RoofStyle
RoofMatl
Exterior1st
Exterior2nd
MasVnrType
ExterQual
ExterCond
Foundation
BsmtQual
BsmtCond
BsmtExposure
BsmtFinType1
BsmtFinType2
Heating
HeatingQC
CentralAir
Electrical
KitchenQual
Functional
FireplaceQu
GarageType
GarageFinish
GarageQual
GarageCond
PavedDrive
PoolQC
Fence
MiscFeature
SaleType
SaleCondition
### Ordinal

### Binary

### Nominal (non-binary)
1. Home Ownership (trinary)
2. Purpose

## Numeric Columns
MSSubClass
MSZoning
LotFrontage
LotArea
Street
Alley
LotShape
LandContour
Utilities
LotConfig
LandSlope
Neighborhood
Condition1
Condition2
BldgType
HouseStyle
OverallQual
OverallCond
YearBuilt
YearRemodAdd
RoofStyle
RoofMatl
Exterior1st
Exterior2nd
MasVnrType
MasVnrArea
ExterQual
ExterCond
Foundation
BsmtQual
BsmtCond
BsmtExposure
BsmtFinType1
BsmtFinSF1
BsmtFinType2
BsmtFinSF2
BsmtUnfSF
TotalBsmtSF
Heating
HeatingQC
CentralAir
Electrical
1stFlrSF
2ndFlrSF
LowQualFinSF
GrLivArea
BsmtFullBath
BsmtHalfBath
FullBath
HalfBath
BedroomAbvGr
KitchenAbvGr
KitchenQual
TotRmsAbvGrd
Functional
Fireplaces
FireplaceQu
GarageType
GarageYrBlt
GarageFinish
GarageCars
GarageArea
GarageQual
GarageCond
PavedDrive
WoodDeckSF
OpenPorchSF
EnclosedPorch
3SsnPorch
ScreenPorch
PoolArea
PoolQC
Fence
MiscFeature
MiscVal
MoSold
YrSold
SaleType
SaleCondition

# Target
SalePrice


In [9]:
# some basic methods to understand the DataFrame
houses.info(); 
display(f"The shape of the DataFrame is {houses.shape}");
display(f"The number of duplicated rows is {houses.duplicated().sum()}");
houses.isna().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

'The shape of the DataFrame is (1460, 81)'

'The number of duplicated rows is 0'

Id                 0
MSSubClass         0
MSZoning           0
LotFrontage      259
LotArea            0
                ... 
MoSold             0
YrSold             0
SaleType           0
SaleCondition      0
SalePrice          0
Length: 81, dtype: int64

In [83]:
houses.isna().sum().sort_values(ascending=False).head(19).index

Index(['PoolQC', 'MiscFeature', 'Alley', 'Fence', 'MasVnrType', 'FireplaceQu',
       'LotFrontage', 'GarageYrBlt', 'GarageCond', 'GarageType',
       'GarageFinish', 'GarageQual', 'BsmtFinType2', 'BsmtExposure',
       'BsmtQual', 'BsmtCond', 'BsmtFinType1', 'MasVnrArea', 'Electrical'],
      dtype='object')

In [68]:
houses[houses.isna().sum().sort_values(ascending=False).head(19).index]

PoolQC MiscFeature Alley  Fence MasVnrType FireplaceQu  LotFrontage  \
0       NaN         NaN   NaN    NaN    BrkFace         NaN         65.0   
1       NaN         NaN   NaN    NaN        NaN          TA         80.0   
2       NaN         NaN   NaN    NaN    BrkFace          TA         68.0   
3       NaN         NaN   NaN    NaN        NaN          Gd         60.0   
4       NaN         NaN   NaN    NaN    BrkFace          TA         84.0   
...     ...         ...   ...    ...        ...         ...          ...   
1455    NaN         NaN   NaN    NaN        NaN          TA         62.0   
1456    NaN         NaN   NaN  MnPrv      Stone          TA         85.0   
1457    NaN        Shed   NaN  GdPrv        NaN          Gd         66.0   
1458    NaN         NaN   NaN    NaN        NaN         NaN         68.0   
1459    NaN         NaN   NaN    NaN        NaN         NaN         75.0   

      GarageYrBlt GarageCond GarageType GarageFinish GarageQual BsmtFinType2  \
0          2003.0         TA     Attchd          RFn         TA          Unf   
1          1976.0         TA     Attchd          RFn         TA          Unf   
2          2001.0         TA     Attchd          RFn         TA          Unf   
3          1998.0         TA     Detchd          Unf         TA          Unf   
4          2000.0         TA     Attchd          RFn         TA          Unf   
...           ...        ...        ...          ...        ...          ...   
1455       1999.0         TA     Attchd          RFn         TA          Unf   
1456       1978.0         TA     Attchd          Unf         TA          Rec   
1457       1941.0         TA     Attchd          RFn         TA          Unf   
1458       1950.0         TA     Attchd          Unf         TA          Rec   
1459       1965.0         TA     Attchd          Fin         TA          LwQ   

     BsmtExposure BsmtQual BsmtCond BsmtFinType1  MasVnrArea Electrical  
0              No       Gd       TA          GLQ       196.0      SBrkr  
1              Gd       Gd       TA          ALQ         0.0      SBrkr  
2              Mn       Gd       TA          GLQ       162.0      SBrkr  
3              No       TA       Gd          ALQ         0.0      SBrkr  
4              Av       Gd       TA          GLQ       350.0      SBrkr  
...           ...      ...      ...          ...         ...        ...  
1455           No       Gd       TA          Unf         0.0      SBrkr  
1456           No       Gd       TA          ALQ       119.0      SBrkr  
1457           No       TA       Gd          GLQ         0.0      SBrkr  
1458           Mn       TA       TA          GLQ         0.0      FuseA  
1459           No       TA       TA          BLQ         0.0      SBrkr  

[1460 rows x 19 columns]

In [89]:
houses[houses.isna().sum().sort_values(ascending=False).head(19).index].count()

PoolQC             7
MiscFeature       54
Alley             91
Fence            281
MasVnrType       588
FireplaceQu      770
LotFrontage     1201
GarageYrBlt     1379
GarageCond      1379
GarageType      1379
GarageFinish    1379
GarageQual      1379
BsmtFinType2    1422
BsmtExposure    1422
BsmtQual        1423
BsmtCond        1423
BsmtFinType1    1423
MasVnrArea      1452
Electrical      1459
dtype: int64

# To do

1. Replace PoolQC NaN with NA where PoolArea = 0
2. Drop MiscFeature since only 54 vals is 0.0369 of data
3. Make Alley binary (0 for NA/NaN, 1 for Grvl,Pave)
4. Make Fence binary: (0 for NA/NaN, 1 for GdPrv,MnPrv,GdWo,MnWw)




In [91]:
for column in houses.columns:
    print(column)

Id
MSSubClass
MSZoning
LotFrontage
LotArea
Street
Alley
LotShape
LandContour
Utilities
LotConfig
LandSlope
Neighborhood
Condition1
Condition2
BldgType
HouseStyle
OverallQual
OverallCond
YearBuilt
YearRemodAdd
RoofStyle
RoofMatl
Exterior1st
Exterior2nd
MasVnrType
MasVnrArea
ExterQual
ExterCond
Foundation
BsmtQual
BsmtCond
BsmtExposure
BsmtFinType1
BsmtFinSF1
BsmtFinType2
BsmtFinSF2
BsmtUnfSF
TotalBsmtSF
Heating
HeatingQC
CentralAir
Electrical
1stFlrSF
2ndFlrSF
LowQualFinSF
GrLivArea
BsmtFullBath
BsmtHalfBath
FullBath
HalfBath
BedroomAbvGr
KitchenAbvGr
KitchenQual
TotRmsAbvGrd
Functional
Fireplaces
FireplaceQu
GarageType
GarageYrBlt
GarageFinish
GarageCars
GarageArea
GarageQual
GarageCond
PavedDrive
WoodDeckSF
OpenPorchSF
EnclosedPorch
3SsnPorch
ScreenPorch
PoolArea
PoolQC
Fence
MiscFeature
MiscVal
MoSold
YrSold
SaleType
SaleCondition
SalePrice


In [84]:
for nancol in houses.isna().sum().sort_values(ascending=False).head(19).index:
    print(f"Value Counts of {houses[nancol].value_counts()}")

Value Counts of PoolQC
Gd    3
Ex    2
Fa    2
Name: count, dtype: int64
Value Counts of MiscFeature
Shed    49
Gar2     2
Othr     2
TenC     1
Name: count, dtype: int64
Value Counts of Alley
Grvl    50
Pave    41
Name: count, dtype: int64
Value Counts of Fence
MnPrv    157
GdPrv     59
GdWo      54
MnWw      11
Name: count, dtype: int64
Value Counts of MasVnrType
BrkFace    445
Stone      128
BrkCmn      15
Name: count, dtype: int64
Value Counts of FireplaceQu
Gd    380
TA    313
Fa     33
Ex     24
Po     20
Name: count, dtype: int64
Value Counts of LotFrontage
60.0     143
70.0      70
80.0      69
50.0      57
75.0      53
        ... 
137.0      1
141.0      1
38.0       1
140.0      1
46.0       1
Name: count, Length: 110, dtype: int64
Value Counts of GarageYrBlt
2005.0    65
2006.0    59
2004.0    53
2003.0    50
2007.0    49
1977.0    35
1998.0    31
1999.0    30
1976.0    29
2008.0    29
2000.0    27
1968.0    26
2002.0    26
1950.0    24
1993.0    22
1958.0    21
1966.0    2